x/gamm/pool-models/balancer/amm.go

```diff
@@ -0,0 +1,143 @@
+package balancer
+
+import (
+	"errors"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/osmosis-labs/osmosis/v7/osmomath"
+	"github.com/osmosis-labs/osmosis/v7/x/gamm/types"
+)
+
+// solveConstantFunctionInvariant solves the constant function of an AMM
+// that determines the relationship between the differences of two sides
+// of assets inside the pool.
+// For fixed balanceXBefore, balanceXAfter, weightX, balanceY, weightY,
+// we could deduce the balanceYDelta, calculated by:
+// balanceYDelta = balanceY * (1 - (balanceXBefore/balanceXAfter)^(weightX/weightY))
+// balanceYDelta is positive when the balance liquidity decreases.
+// balanceYDelta is negative when the balance liquidity increases.
+func solveConstantFunctionInvariant(
+	tokenBalanceFixedBefore,
+	tokenBalanceFixedAfter,
+	tokenWeightFixed,
+	tokenBalanceUnknownBefore,
+	tokenWeightUnknown sdk.Dec,
+) sdk.Dec {
+	// weightRatio = (weightX/weightY)
+	weightRatio := tokenWeightFixed.Quo(tokenWeightUnknown)
+
+	// y = balanceXBefore/balanceYAfter
+	y := tokenBalanceFixedBefore.Quo(tokenBalanceFixedAfter)
+
+	// amountY = balanceY * (1 - (y ^ weightRatio))
+	foo := osmomath.Pow(y, weightRatio)
+	multiplier := sdk.OneDec().Sub(foo)
+	return tokenBalanceUnknownBefore.Mul(multiplier)
+}
+
+func (p Pool) parsePoolAssetsByDenoms(tokenADenom, tokenBDenom string) (
+	Aasset types.PoolAsset, Basset types.PoolAsset, err error) {
+	Aasset, found1 := types.GetPoolAssetByDenom(p.PoolAssets, tokenADenom)
+	Basset, found2 := types.GetPoolAssetByDenom(p.PoolAssets, tokenBDenom)
+	if !(found1 && found2) {
+		return Aasset, Basset, errors.New("TODO: fill message here")
+	}
+	return Aasset, Basset, nil
+}
+
+func (p Pool) parsePoolAssets(tokensA sdk.Coins, tokenBDenom string) (
+	tokenA sdk.Coin, Aasset types.PoolAsset, Basset types.PoolAsset, err error) {
+	if len(tokensA) != 1 {
+		return tokenA, Aasset, Basset, errors.New("TODO: Fill message here")
+	}
+	Aasset, Basset, err = p.parsePoolAssetsByDenoms(tokensA[0].Denom, tokenBDenom)
+	return tokensA[0], Aasset, Basset, err
+}
+
+// CalcOutAmtGivenIn calculates token to be swapped out given
```

Comment:
I love this approach where poolAssets are no longer poolI but went in balancerPoolAsset